Generic controller

SomtochiAma opened the pull request.

dholbach commented:
/ok-to-test
/uncc
/cc @stealthybox

justinsb commented on generic/main.go:
```diff
@@ -0,0 +1,131 @@
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = addonsv1alpha1.AddToScheme(scheme) // <--------
```
Nit: We can remove the `<---` marker :-)

justinsb commented on generic/main.go:
```diff
@@ -0,0 +1,131 @@
+		os.Exit(1)
+	}
+
+	for i := range genericList.Items {
```
Nit: I think we should have a TODO to say "add a watch on Generic, respond to changes" or something like that

justinsb commented on generic/controllers/generic_controller.go:
```diff
@@ -0,0 +1,92 @@
+	Channel string
+}
+
+// +kubebuilder:rbac:groups=addons.x-k8s.io,resources=generics,verbs=get;list;watch;create;update;patch;delete
```
I actually don't think we need these rbac permissions.  I guess we want get/list/watch on generics, but you could also move than to main.go.

SomtochiAma replied:
Is there any difference, moving it to `main.go`

justinsb commented on generic/controllers/generic_controller.go:
```diff
@@ -0,0 +1,92 @@
+		return err
+	}
+
+	// Watch for changes to Generic
```
Nit: I think this should say "Watch for changes to the objectKind" or something like that

justinsb commented:
This is looking really good!  We do need the tools to generate the RBAC & the manifests, but then this is going to be a really nice experience for users!

cmoulliard commented:
Can you please add a `README.md` file under the tools in order to explain how to use them with some concrete examples ?

cmoulliard commented on generic/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+FROM gcr.io/distroless/static:nonroot
+WORKDIR /
+COPY --from=builder /workspace/manager .
+COPY addons/ addons/
```
Error. The line `COPY addons/ addons/` should be changed to ` COPY channels/ channels/` otherwise docker build will fail.

cmoulliard commented:
I got this error when I run the generic controller under a kind cluster using the command `make deploy`

```
kc logs generic-controller-manager-5d9cf5db7b-mnfc2 -n generic-system -c manager
2020-08-06T10:27:56.581Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
Please create a `Generic` resource
```
Why should we do then deploy first a Generic CRD before to start the controler ?

SomtochiAma commented:
> I got this error when I run the generic controller under a kind cluster using the command `make deploy`
> 
> ```
> kc logs generic-controller-manager-5d9cf5db7b-mnfc2 -n generic-system -c manager
> 2020-08-06T10:27:56.581Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
> Please create a `Generic` resource
> ```
> 
> Why should we do then deploy first a Generic CRD before to start the controler ?

The generic controller is different because it is used as a general controller for other simple addons that don't require any extra configuration. 
The Generic resource is used to tell the generic controller these addons are and where to find their channels

SomtochiAma commented:
> Can you please add a `README.md` file under the tools in order to explain how to use them with some concrete examples ?

I will definitely do so

cmoulliard commented:
> The Generic resource is used to tell the generic controller these addons are and where to find their channels

Maybe then review the message logged to make it more explicit about root cause of the error and what is it needed to be done to avoid it. As the controller will crash, then we must find a way to make it healthy if a Generic CR has been deployed ...

SomtochiAma commented:
> Maybe then review the message logged to make it more explicit about root cause of the error and what is it needed to be done to avoid it. 

I changed the message and added a REAME.md for the tools. Please review whenever you can! Thanks

justinsb commented:
This is really cool and it sounds like people want to start trying it out, so let's merge it and iterate with further improvements!

/approve
/lgtm
